source/service.go

```diff
@@ -43,37 +43,56 @@ type ServiceSource struct {
 
 // Endpoints returns endpoint objects for each service that should be processed.
 func (sc *ServiceSource) Endpoints() ([]endpoint.Endpoint, error) {
-	services, err := sc.Client.Core().Services(v1.NamespaceAll).List(v1.ListOptions{})
+	services, err := sc.Client.CoreV1().Services(v1.NamespaceAll).List(v1.ListOptions{})
 	if err != nil {
 		return nil, err
 	}
 
 	endpoints := []endpoint.Endpoint{}
 
 	for _, svc := range services.Items {
-		// Check controller annotation to see if we are responsible.
-		controller, exists := svc.Annotations[controllerAnnotationKey]
-		if exists && controller != controllerAnnotationValue {
-			continue
+		svcEndpoints := endpointsFromService(&svc)
+		if len(svcEndpoints) != 0 {
+			endpoints = append(endpoints, svcEndpoints...)
 		}
+	}
 
-		// Get the desired hostname of the service from the annotation.
-		hostname, exists := svc.Annotations[hostnameAnnotationKey]
-		if !exists {
-			continue
-		}
+	return endpoints, nil
+}
 
-		// Create an endpoint matching the desired hostname.
-		endpoint := endpoint.Endpoint{
-			DNSName: hostname,
-		}
+// endpointsFromService extracts the endpoints from a service object
+func endpointsFromService(svc *v1.Service) []endpoint.Endpoint {
+	var endpoints []endpoint.Endpoint
+
+	// Check controller annotation to see if we are responsible.
+	controller, exists := svc.Annotations[controllerAnnotationKey]
+	if exists && controller != controllerAnnotationValue {
+		return endpoints
+	}
+
+	// Get the desired hostname of the service from the annotation.
+	hostname, exists := svc.Annotations[hostnameAnnotationKey]
+	if !exists {
+		return endpoints
+	}
 
-		// Create a corresponding endpoint for each configured external entrypoint.
-		for _, lb := range svc.Status.LoadBalancer.Ingress {
-			endpoint.Target = lb.IP
+	// Create a corresponding endpoint for each configured external entrypoint.
+	for _, lb := range svc.Status.LoadBalancer.Ingress {
+		if lb.IP != "" {
+			endpoint := endpoint.Endpoint{
+				DNSName: hostname,
+				Target:  lb.IP,
```

Comment:
i would refactor this to less lines: 

endpoint := endpoint.Endpoint{
DNSName: hostname
}
//populate target based on lb.IP value